if_perl.xs: fix compatibility for Perl 5.36 and up

pheiduck opened the pull request:
Test Env:
Fedora 39 + Perl 5.37.11 and Perl 5.38.0-RC1
`./configure --enable-gui=no --with-features=Huge --enable-perlinterp`

Fixes: https://github.com/vim/vim/issues/12543
Closes: https://github.com/Perl/perl5/issues/21164

pheiduck committed:
if_perl.xs: define version for a workaround

pheiduck committed:
fixup: more build errors with Perl 5.37.x

pheiduck committed:
CI: use latest Perl provided by homebrew

pheiduck committed:
use --enable-perlinterp=dynamic

It's needed on newer Perl Versions, static not working

pheiduck committed:
avoid "-Werror" and "-Wshadow" on macOS

pheiduck committed:
Remove unknown "pragma GCC diagnostic ignored"

pheiduck committed:
use --enable-perlinterp again

should work now

ichizok commented on src/if_perl.xs:
```diff
@@ -56,6 +56,8 @@
 # pragma GCC diagnostic push
 # pragma GCC diagnostic ignored "-Wunused-variable"
+# pragma GCC diagnostic ignored "-Wincompatible-pointer-types"
+# pragma GCC diagnostic ignored "-Wshadow"
```
I think should not ignore these warnings (`-Wincompatible-pointer-types` and `-Wshadow`) by pragma.
For CI, should add them to `PERL_CFLAGS_EXTRA` in `ci/config.mk.sed`.

pheiduck replied:
done: https://github.com/vim/vim/pull/12575/commits/6abf8cdab0bf03ea05cec19d538d4387d730a933

k-takata commented on src/if_perl.xs:
```diff
@@ -40,7 +40,7 @@
  * Don't include "perl\lib\CORE\inline.h" for now,
  * include it after Perl_sv_free2 is defined. */
-#ifdef DYNAMIC_PERL
+#if (PERL_REVISION == 5) && (PERL_VERSION == 18)
```
This comment says "work around for perl-5.18", but I think actually this is a workaround for perl-5.18 and later.
Not sure which versions are relevant, though.

pheiduck replied:
done: https://github.com/vim/vim/pull/12575/commits/6abf8cdab0bf03ea05cec19d538d4387d730a933

pheiduck committed:
fixup: move to right place PERL_CFLAGS; workaround for perl-5.18 and later.

pheiduck committed:
config.mk.sed: use -Wno-shadow

pheiduck committed:
apply suggested patch by @zdohnal

pheiduck committed:
fixup: use GIMME_V instead

It is planned to remove GIMME from a future release of Perl. Do not use it for new code; remove it from existing code.
